### PR TITLE
nixos/prometheus: add `enableAgentMode` option

### PR DESCRIPTION
**Pull request by Drakonis**
###### Description of changes
Add `enableAgentMode` option to prometheus service.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by otavio — APPROVED** [review]

File: nixos/modules/services/monitoring/prometheus/default.nix [hunk 1]
```diff
@@ -41,12 +41,12 @@ let
   # This becomes the main config file for Prometheus
   promConfig = {
     global = filterValidPrometheus cfg.globalConfig;
-    rule_files = map (promtoolCheck "check rules" "rules") (cfg.ruleFiles ++ [
-      (pkgs.writeText "prometheus.rules" (concatStringsSep "\n" cfg.rules))
-    ]);
     scrape_configs = filterValidPrometheus cfg.scrapeConfigs;
     remote_write = filterValidPrometheus cfg.remoteWrite;
     remote_read = filterValidPrometheus cfg.remoteRead;
+    rule_files = optionals (cfg.enableAgentMode != true) (map (promtoolCheck "check rules" "rules") (cfg.ruleFiles ++ [
```
Comment (AndersonTorres) [inline nixos/modules/services/monitoring/prometheus/default.nix:47]:
```suggestion
    rule_files = optionals (!(cfg.enableAgentMode)) (map (promtoolCheck "check rules" "rules") (cfg.ruleFiles ++ [
```

`X!=true` is the same as
`!X`.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Review by otavio — CHANGES_REQUESTED** [review]

**Review by otavio — APPROVED** [review]

**Review by thiagokokada — APPROVED** [review]

File: nixos/modules/services/monitoring/prometheus/default.nix [hunk 2]
```diff
@@ -62,15 +62,16 @@ let
...
       if cfg.enableReload
       then "/etc/prometheus/prometheus.yaml"
       else prometheusYml
     }"
     "--web.listen-address=${cfg.listenAddress}:${builtins.toString cfg.port}"
-    "--alertmanager.notification-queue-capacity=${toString cfg.alertmanagerNotificationQueueCapacity}"
-  ] ++ optional (cfg.webExternalUrl != null) "--web.external-url=${cfg.webExternalUrl}"
+  ] ++ optional (!(cfg.enableAgentMode)) "--alertmanager.notification-queue-capacity=${toString cfg.alertmanagerNotificationQueueCapacity }"
+    ++ optional (!(cfg.enableAgentMode)) "--storage.tsdb.path=${workingDir}/data/"
+    ++ optional (cfg.enableAgentMode) "--enable-feature=agent"
```
Comment (AndersonTorres) [inline nixos/modules/services/monitoring/prometheus/default.nix:73]:
```suggestion
  ] ++ ( 
    if (cfg.enableAgentMode) then [
      "--enable-feature=agent" 
    ] else [ 
       "--alertmanager.notification-queue-capacity=${toString cfg.alertmanagerNotificationQueueCapacity }" 
       "--storage.tsdb.path=${workingDir}/data/"
    ])
```

(Ignore my indentation)

File: nixos/modules/services/monitoring/prometheus/default.nix [hunk 3]
```diff
@@ -1612,6 +1613,14 @@ in
       '';
     };
 
+    enableAgentMode = mkOption {
+      default = false;
+      type = types.bool;
+      description = lib.mdDoc ''
```
Comment (AndersonTorres) [inline nixos/modules/services/monitoring/prometheus/default.nix:1619]: 🙄

Ignoring.

**Review by otavio — APPROVED** [review]

File: nixos/modules/services/monitoring/prometheus/default.nix [hunk 4]
```diff
@@ -1612,6 +1617,14 @@ in
       '';
     };
 
+    enableAgentMode = mkOption {
```
Comment (thiagokokada) [inline nixos/modules/services/monitoring/prometheus/default.nix:1620]: `mkEnableOption` ?

File: nixos/modules/services/monitoring/prometheus/default.nix [hunk 5]
```diff
@@ -1612,6 +1617,8 @@ in
       '';
     };
 
+    enableAgentMode = mkEnableOption (lib.mdDoc "Enables agent mode.");
```
Comment (thiagokokada) [inline nixos/modules/services/monitoring/prometheus/default.nix:1620]: `mkEnableOption` starts the phrase with `Whether to enable` and ends with a period already, so the only think you need is `agent mode` here.

**Review by thiagokokada — APPROVED** [review]

**Pull request merged**